README.md
cabal run websocket-server

or
cabal install --overwrite-policy=always
websocket-server




## Notes
I think I need to bite the bullet and have the server return a webpage with SSE frontend which will make it's own connection


CHANGELOG.md
# Revision history for websocket-server

## 0.1.0.0 -- YYYY-mm-dd

* First version. Released on an unsuspecting world.


app/Test.hs
{-# LANGUAGE OverloadedStrings #-}

module Test (xxx) where

import Data.Attoparsec.ByteString.Char8 (parseOnly)
import Data.ByteString (ByteString)
import Parser
import Syntax

req :: ByteString
req = "GET /user-agent HTTP/1.1\r\n\r\nHost: localhost:4221\r\nUser-Agent: curl/7.64.1\r\n"

req2 :: ByteString
req2 = "GET /echo/hello HTTP/1.1\r\nHost: 127.0.0.1:4221\r\nUser-Agent: curl/7.88.1\r\nAccept: */*\r\n\r\n"

xxx :: Either String Req
xxx = parseOnly parseReq req2


app/Main.hs
{-# LANGUAGE OverloadedStrings #-}

module Main (main) where

import Control.Monad (forever)
import Data.ByteString (ByteString)
import qualified Data.ByteString as B
import Data.ByteString.Char8 (pack)
import qualified Data.ByteString.Lazy.Char8 as BLC
import qualified Format
import Handle
import Parser (parseReq, parseOnly)

--import Network.Simple.TCP (HostPreference (..), recv, send, serve)
import Syntax
import System.Environment (getArgs)

import Control.Concurrent (ThreadId, forkIO, threadDelay)

--import Data.Function
--import Data.Functor
import Network.Socket (
    Family (..),
    PortNumber (..),
    SockAddr (..),
    Socket,
    SocketOption (..),
    SocketType (..),
    accept,
    bind,
    close,
    defaultProtocol,
    listen,
    setSocketOption,
    socket,
 )
import Network.Socket.ByteString (recv, send, sendAll, sendAllTo, sendTo)

did :: ByteString
did = "id: 0"

ddata :: ByteString
ddata = "data: hello world"

datum :: ByteString
datum = "data: hello world"

hi :: ByteString
hi = "8data: hi\r\n"

sser :: ByteString
--sser = "200 OK\r\ncontent-type: text/event-stream\r\n\r\n"
sser = "HTTP/1.1 200 OK\r\nContent-Type: text/event-stream\r\nConnection: Keep-Alive\r\nContent-Length: 0\r\n\r\n"

main :: IO ()
main = do
    argv <- getArgs

    BLC.putStrLn "Logs from your program will appear here"

    let host = 1024
        port = 4221 :: PortNumber
        -- TODO have env include host and port
        env = case argv of
            ["--directory", directory] -> Env $ pack directory
            _ -> Env B.empty

    BLC.putStrLn $ "Listening on port " <> BLC.pack (show port)

    sock <- socket AF_INET Stream defaultProtocol
    setSocketOption sock ReuseAddr 1
    bind sock (SockAddrInet port 0)
    listen sock 5
    forever $ do
        (conn, _address) <- accept sock
        bReq <- recv conn host
        forkIO $ case parseOnly parseReq bReq of
            Left _ -> close conn
            Right req -> handle env conn req


app/Format.hs
{-# LANGUAGE NamedFieldPuns #-}
{-# LANGUAGE OverloadedStrings #-}

module Format where

import Data.ByteString (ByteString)
import qualified Data.ByteString as B
import Syntax

class Pack a where
    pack :: a -> ByteString

instance Pack Protocol where
    pack HTTP1_0 = "HTTP/1.0"
    pack HTTP1_1 = "HTTP/1.1"
    pack HTTP2_0 = "HTTP/2.0"

instance Pack Status where
    pack (Status 200) = "200 OK"
    pack (Status 201) = "201 CREATED"
    pack (Status _) = "404 NOT FOUND"

instance Pack Resp where
    pack Resp{protocol', status, headers', body'} =
        pack protocol'
            `B.append` " "
            `B.append` pack status
            `B.append` "\r\n"
            `B.append` B.intercalate "\r\n" ((\(k, v) -> B.concat [k, ": ", v]) <$> headers')
            `B.append` "\r\n\r\n"
            `B.append` body'


app/Parser.hs
{-# LANGUAGE LambdaCase #-}
{-# LANGUAGE OverloadedStrings #-}

module Parser (parseReq, parseOnly) where

import Syntax

import Control.Applicative
import Data.Attoparsec.ByteString.Char8 (Parser, endOfInput, endOfLine, isSpace, many', parseOnly, skipSpace, space, string, take, takeTill, takeByteString)
import Data.ByteString (ByteString)
import Data.ByteString.Char8 (readInt)
import Data.Function
import Data.Functor
import Prelude hiding (take)

parseMethod :: Parser Method
parseMethod =
    (string "GET" $> GET)
        <|> (string "POST" $> POST)

parsePath :: Parser Path
parsePath = Path <$> takeTill isSpace

data Route = Whack | Agent | Sse | File ByteString | Html ByteString | Echo ByteString

parseRoute :: Parser Route
parseRoute = (string "/" *> endOfInput $> Whack)
        <|> (string "/user-agent" *> endOfInput $> Agent)
        <|> (string "/sse" $> Sse)
        <|> (string "/files/" *> takeByteString <&> File)
        <|> (string "/html/" *> takeByteString <&> Html)
        <|> (string "/echo/" *> takeByteString <&> Echo)

parseProtocol :: Parser Protocol
parseProtocol =
    (string "HTTP/1.0" $> HTTP1_0)
        <|> (string "HTTP/1.1" $> HTTP1_1)
        <|> (string "HTTP/2.0" $> HTTP2_0)

parseLine :: Parser ByteString
parseLine = takeTill (== '\r') <* endOfLine

parseHeader :: Parser KeyVal
parseHeader = liftA2 (,) (takeTill (== ':') <* string ":" <* skipSpace) parseLine

parseHeaders :: Parser Map
parseHeaders = many' parseHeader

headersToLength :: Map -> Int
headersToLength hds =
    hds
        & getHeader "Content-Length"
        & readInt
        & \case
            Just (int, _) -> int
            Nothing -> 0

parseHeadersAndBody :: Parser (Map, Body)
parseHeadersAndBody = parseHeaders >>= \h -> endOfLine *> (take . headersToLength) h <&> \b -> (h, Body b)

parseReq :: Parser Req
parseReq =
    (\m pa pr (he, bo) -> Req m pa pr he bo) <$> parseMethod
        <* space
        <*> parsePath
        <* space
        <*> parseProtocol
        <* endOfLine
        <*> parseHeadersAndBody


app/Syntax.hs
module Syntax where

import Data.ByteString (ByteString)
import qualified Data.ByteString as B
import Data.Function
import Data.Maybe (fromMaybe)
--import Data.List (lookup)

type KeyVal = (ByteString, ByteString)
type Map = [KeyVal]

getHeader :: ByteString -> Map -> ByteString
getHeader key hs = lookup key hs & fromMaybe B.empty

newtype Env = Env { dir :: ByteString }

newtype Path = Path ByteString
    deriving (Eq, Show)

newtype Body = Body ByteString
    deriving (Eq, Show)

newtype Status = Status Integer
    deriving (Eq, Show)

data Method = GET | POST
    deriving (Eq, Show)

data Protocol = HTTP1_0 | HTTP1_1 | HTTP2_0
    deriving (Eq, Show)

data Req = Req
    { method :: Method
    , path :: Path
    , protocol :: Protocol
    , headers :: Map
    , body :: Body
    }
    deriving (Eq, Show)

data Resp = Resp
    { protocol' :: Protocol
    , status :: Status
    , headers' :: Map
    , body' :: ByteString
    }
    deriving (Eq, Show)


app/ParserExample.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE TupleSections #-}

module ParserExample where

import Data.Char (digitToInt)
import Control.Applicative
import Data.Attoparsec.ByteString.Char8 ( Parser, parseOnly, string, take, digit)
import Data.ByteString (ByteString)
import Data.Functor
import Prelude hiding (take)

data Version = One | Two
data Protocol' = Protocol' { version :: Version, length :: Int, body :: ByteString }

parseVersion :: Parser Version
parseVersion = (string "1" $> One) <|> (string "2" $> Two)

parseBody :: Parser (Int, ByteString)
parseBody = digit <&> digitToInt >>= \len -> take len <&> (len,)

-- HELP! can parseProtocol' be rewritten without the lambda function??
parseProtocol' :: Parser Protocol'
--parseProtocol' = (\v (l, b) -> Protocol' v l b) <$> parseVersion <*> parseBody
parseProtocol' = uncurry . Protocol' <$> parseVersion <*> parseBody


data Protocol2 = Protocol2 { l :: Int, b :: ByteString }

parseLength :: Parser Int
parseLength = digitToInt <$> digit

parseBody' :: Int -> Parser ByteString
parseBody' = take

parseProtocol2 :: Parser Protocol2
parseProtocol2 = Protocol2 <$> parseLength >>= \p -> p <$> parseBody' 3


app/Handle.hs
{-# LANGUAGE LambdaCase #-}
{-# LANGUAGE NamedFieldPuns #-}
{-# LANGUAGE OverloadedStrings #-}

module Handle (handle) where

import Control.Applicative
import Data.ByteString (ByteString)
import qualified Data.ByteString as B
import Data.ByteString.Char8 (pack)
import Data.Functor
import System.Directory (doesFileExist)
import Network.Socket ( Socket)

--import Data.Attoparsec.ByteString.Char8 (Parser, char8, count, decimal, digit, endOfLine, isSpace, parseOnly, skipSpace, space, string, take, takeTill)
import Data.Attoparsec.ByteString.Char8 (Parser, endOfInput, parseOnly, string, takeByteString)
import qualified Data.ByteString as B
import qualified Format
import Parser
import Syntax

toHeader :: ByteString -> ByteString -> (ByteString, ByteString)
toHeader = (,)

ok :: ByteString -> Integer -> ByteString -> Resp
ok contentType status body' =
    Resp
        { protocol' = HTTP1_1
        , status = Status status
        , headers' =
            [ toHeader "Content-Type" contentType
            , toHeader "Content-Length" ((pack . show . B.length) body')
            ]
        , body'
        }

sse :: Resp
sse = ok "text/event-stream" 200 "data: hi"

txt :: ByteString -> Resp
txt = ok "text/plain" 200

file :: ByteString -> Resp
file = ok "application/octet-stream" 200

write :: Resp
write = ok "text/plain" 201 B.empty

html :: ByteString -> Resp
html = ok "text/html" 200

notFound :: Resp
notFound =
    Resp
        { protocol' = HTTP1_1
        , status = Status 404
        , headers' = []
        , body' = ""
        }

parseRoute :: Parser [ByteString]
parseRoute =
    (string "/" *> endOfInput $> ["/"])
        <|> (string "/user-agent" *> endOfInput $> ["user-agent"])
        <|> (string "/files/" *> takeByteString <&> \path -> ["files", path])
        <|> (string "/html/" *> takeByteString <&> \path -> ["html", path])
        <|> (string "/sse" $> ["sse"])
        <|> (string "/echo/" *> takeByteString <&> \echo -> ["echo", echo])

routeToResp :: Env -> Method -> Req -> [ByteString] -> IO Resp
routeToResp _ GET _ ["/"] = pure $ txt ""
routeToResp _ GET _ ["echo", echo] = pure $ txt echo
routeToResp _ GET _ ["sse"] = pure sse
routeToResp _ GET Req{headers} ["user-agent"] = pure $ txt (getHeader "User-Agent" headers)
routeToResp Env{dir} GET _ ["html", bsPath] =
    B.toFilePath (dir <> "/" <> bsPath)
        >>= doesFileExist
        >>= \exists ->
            if exists
                then B.toFilePath (dir <> "/" <> bsPath) >>= B.readFile <&> html
                else pure notFound
routeToResp Env{dir} GET _ ["files", bsPath] =
    B.toFilePath (dir <> "/" <> bsPath)
        >>= doesFileExist
        >>= \exists ->
            if exists
                then B.toFilePath (dir <> "/" <> bsPath) >>= B.readFile <&> file
                else pure notFound
routeToResp Env{dir} POST Req{body = Body body} ["files", bsPath] = do
    path <- B.toFilePath (dir <> "/" <> bsPath)
    B.writeFile path body
    pure write
routeToResp _ _ _ _ = pure notFound


handle :: Env -> Socket -> Req -> IO ()
handle env conn req@Req{path = (Path path), method} =
    case parseOnly parseRoute path of
        Right bs -> routeToResp env method req bs
        Left _ -> pure notFound


onRequest :: Env -> Socket -> Req -> IO ()
onRequest env conn req = undefined
    --res <- handle env req
    --let isSse = B.empty == getHeader "text/event-stream" (headers' res)
    --if isSse
        --then do
            --_ <- print "A"
            --_ <- print (Format.pack res)
            --_ <- sendTo conn (Format.pack res) addr
            --_ <- sendTo conn hi addr
            --return ()
        --else --close conn
        --do
            --_ <- print "B"
            --_ <- send conn (Format.pack res)
            --close conn
